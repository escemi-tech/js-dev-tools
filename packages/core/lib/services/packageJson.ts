import { existsSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";

import deepmerge from "deepmerge";
import { PackageJson as PackageJsonType } from "type-fest";

export function getPackageJsonPath(dirPath: string) {
  const packageJsonPath = join(dirPath, "package.json");

  if (!existsSync(packageJsonPath)) {
    throw new Error(`No package.json found in directory "${dirPath}"`);
  }

  return packageJsonPath;
}

export type PackageJson = PackageJsonType & {
  [key: string]: any;
};

export function readPackageJson(dirPath: string): PackageJson {
  return JSON.parse(readFileSync(getPackageJsonPath(dirPath), "utf-8")) as PackageJson;
}

export function updatePackageJson(dirPath: string, data: any) {
  const packageJsonPath = getPackageJsonPath(dirPath);
  const originalData = JSON.parse(readFileSync(packageJsonPath, { encoding: "utf8" }));
  const newData = deepmerge(originalData, data);

  writeFileSync(packageJsonPath, JSON.stringify(newData));
}
