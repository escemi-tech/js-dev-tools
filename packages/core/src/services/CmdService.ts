import { spawn } from "child_process";
import { existsSync, statSync } from "fs";

export class CmdService {
  static execCmd(args: string | string[], cwd?: string, silent = false): Promise<string> {
    if (!args.length) {
      throw new Error("Command args must not be empty");
    }

    if (cwd) {
      if (!existsSync(cwd) || !statSync(cwd).isDirectory()) {
        throw new Error(`Directory "${cwd}" does not exist`);
      }
    }

    let cmd: string;
    if (Array.isArray(args)) {
      cmd = args.shift() || "";
    } else {
      cmd = args;
      args = [];
    }

    return new Promise((resolve, reject) => {
      const child = spawn(cmd, args as string[], {
        stdio: silent ? "pipe" : "inherit",
        shell: true,
        windowsVerbatimArguments: true,
        cwd,
      });

      let output = "";
      let error = "";

      child.on("exit", function (code) {
        if (code) {
          return reject(error);
        }
        resolve(output);
      });

      if (child.stdout) {
        child.stdout.on("data", (data) => {
          output += `\n${data}`;
        });
      }
      if (child.stderr) {
        child.stderr.on("data", (data) => {
          error += `\n${data}`;
        });
      }
    });
  }
}
