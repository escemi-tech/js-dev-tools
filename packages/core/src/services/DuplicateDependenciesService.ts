import { resolve } from "path";

import { PackageJson } from "./PackageJson";

export class DuplicateDependenciesService {
  static duplicateDependencies(tsDevToolsRootPath: string, absoluteProjectDir: string): void {
    const projectPackageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const installedPlugins = projectPackageJson.getInstalledPlugins();

    console.info(`Checking for duplicate dev dependencies...`);

    for (const plugin of installedPlugins) {
      DuplicateDependenciesService.duplicatePluginDependencies(
        tsDevToolsRootPath,
        projectPackageJson,
        plugin
      );
    }

    console.info(`Check for duplicate dev dependencies done!`);
  }

  private static duplicatePluginDependencies(
    tsDevToolsRootPath: string,
    projectPackageJson: PackageJson,
    plugin: string
  ): void {
    const absolutePluginDir = resolve(tsDevToolsRootPath, "../../", plugin);
    const pluginPackageJson = PackageJson.fromDirPath(absolutePluginDir);

    const projectDevDependencies = projectPackageJson.getDevDependenciesPackageNames();
    const pluginDependencies = pluginPackageJson.getDependenciesPackageNames();

    const duplicateDependencies = projectDevDependencies.filter((dependency) =>
      pluginDependencies.includes(dependency)
    );

    if (duplicateDependencies.length) {
      console.info(
        `Some dev dependencies are unnecessarily installed as their are already required by "${plugin}":"\n  - "${duplicateDependencies.join(
          "\n  - "
        )}\n`
      );
    }
  }
}
