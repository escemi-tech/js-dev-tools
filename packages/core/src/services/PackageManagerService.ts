import { existsSync } from "fs";
import { join } from "path";

export enum PackageManagerType {
  yarn = "yarn",
  npm = "npm",
}

export class PackageManagerService {
  static detectPackageManager(dirPath: string): PackageManagerType {
    if (existsSync(join(dirPath, "yarn.lock"))) {
      return PackageManagerType.yarn;
    }
    return PackageManagerType.npm;
  }
}
