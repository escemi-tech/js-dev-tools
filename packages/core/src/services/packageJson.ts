import { copyFileSync, existsSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";

import deepmerge from "deepmerge";
import { PackageJson as PackageJsonType } from "type-fest";

export function getPackageJsonPath(dirPath: string): string {
  const packageJsonPath = join(dirPath, "package.json");

  if (!existsSync(packageJsonPath)) {
    throw new Error(`No package.json found in directory "${dirPath}"`);
  }

  return packageJsonPath;
}

export type JsonArray = boolean[] | number[] | string[] | JsonFileData[] | Date[];
type AnyJson = boolean | number | string | JsonFileData | Date | JsonArray | JsonArray[];

export type JsonFileData = {
  [key: string]: AnyJson | undefined;
};
export type PackageJson = JsonFileData & PackageJsonType;

export function readPackageJson(dirPath: string): PackageJson {
  return JSON.parse(readFileSync(getPackageJsonPath(dirPath), "utf-8")) as PackageJson;
}

export function updatePackageJson(dirPath: string, data: Partial<PackageJson>): void {
  const packageJsonPath = getPackageJsonPath(dirPath);
  const originalData = JSON.parse(readFileSync(packageJsonPath, { encoding: "utf8" }));
  const newData = deepmerge(originalData, data, {
    // When merging array, return unique values if array is containingg string only
    arrayMerge: (target, source) => {
      const merged = [...source, ...target];

      const containsOnlyString = merged.every((value) => typeof value === "string");
      if (!containsOnlyString) {
        return merged;
      }

      return merged.filter((v, i, a) => a.indexOf(v) === i);
    },
  });

  writeFileSync(packageJsonPath, JSON.stringify(newData, null, 2));
}

export function getInstalledPlugins(dirPath: string): string[] {
  const packageJson = readPackageJson(dirPath);

  const devDependencies = packageJson.devDependencies;
  if (!devDependencies) {
    return [];
  }

  return Object.keys(devDependencies).filter(
    (devDependency) =>
      devDependency.match(/^@ts-dev-tools\/.*$/) && devDependency !== "@ts-dev-tools/core"
  );
}

export function backupPackageJson(absoluteProjectDir: string): string {
  // Backup package.json
  const packageJsonPath = getPackageJsonPath(absoluteProjectDir);
  const packageJsonBackupPath = packageJsonPath + ".backup";
  copyFileSync(packageJsonPath, packageJsonBackupPath);
  return packageJsonBackupPath;
}

export function revertPackageJson(absoluteProjectDir: string, packageJsonBackupPath: string): void {
  const packageJsonPath = getPackageJsonPath(absoluteProjectDir);
  copyFileSync(packageJsonBackupPath, packageJsonPath);
}
