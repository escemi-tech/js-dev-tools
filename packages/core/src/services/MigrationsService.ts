import { readdirSync, unlinkSync } from "fs";
import { resolve } from "path";

import { PackageJson } from "../services/PackageJson";

export type Migration = { name: string; path: string };

export type MigrationUpFunction = (absoluteProjectDir: string) => Promise<void>;

export class MigrationsService {
  static async executeMigrations(
    tsDevToolsRootPath: string,
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Promise<void> {
    const migrations = MigrationsService.getAvailableMigrations(
      tsDevToolsRootPath,
      absoluteProjectDir,
      currentVersion
    );

    const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const packageJsonBackupPath = packageJson.backup();

    try {
      for (const migration of migrations) {
        console.info(`Applying migration "${migration.name}"...`);

        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { up } = require(migration.path);

        // Apply migration
        await up(absoluteProjectDir);

        // Upgrade current version
        PackageJson.fromDirPath(absoluteProjectDir).merge({
          tsDevTools: { version: migration.name },
        });

        console.info(`Migration "${migration.name}" applied!`);
      }
    } catch (error) {
      // Rollback package.json
      packageJson.restore(packageJsonBackupPath);
      throw error;
    }

    unlinkSync(packageJsonBackupPath);
  }

  private static getAvailableMigrations(
    tsDevToolsRootPath: string,
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Migration[] {
    const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const installedPlugins = packageJson.getInstalledPlugins();

    const migrationFiles: Migration[] = [];
    for (const installedPlugin of installedPlugins) {
      migrationFiles.push(
        ...MigrationsService.getPluginMigrations(
          tsDevToolsRootPath,
          installedPlugin,
          currentVersion
        )
      );
    }

    migrationFiles.sort(({ name: nameA }, { name: nameB }) => nameA.localeCompare(nameB));

    return Array.from(new Set(migrationFiles));
  }

  private static getPluginMigrations(
    tsDevToolsRootPath: string,
    plugin: string,
    currentVersion: string | undefined
  ): Migration[] {
    const pluginDirPath = resolve(tsDevToolsRootPath, "../../", plugin);

    // First retrieve migration of inherited plugins
    const migrationFiles = MigrationsService.getAvailableMigrations(
      tsDevToolsRootPath,
      pluginDirPath,
      currentVersion
    );

    // Then retrieve
    const pluginMigrationsDirPath = resolve(pluginDirPath, "dist/install/migrations");
    for (const migrationFile of readdirSync(pluginMigrationsDirPath)) {
      if (!migrationFile.match(/^[0-9]{14}-[a-z]+\.(js|ts)$/)) {
        continue;
      }

      const migrationName = migrationFile.split(".").slice(0, -1).join(".");
      const shouldApplyMigration =
        !currentVersion || currentVersion.localeCompare(migrationName) < 0;

      if (!shouldApplyMigration) {
        continue;
      }

      migrationFiles.push({
        name: migrationName,
        path: resolve(pluginMigrationsDirPath, migrationFile),
      });
    }
    return migrationFiles;
  }
}
