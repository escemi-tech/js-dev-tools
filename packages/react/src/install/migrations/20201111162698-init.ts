import { PackageJson } from "@ts-dev-tools/core/dist/services/PackageJson";
import { PackageJsonMerge } from "@ts-dev-tools/core/dist/services/PackageJsonMerge";

export async function up(absoluteProjectDir: string): Promise<void> {
  const eslintConfig = {
    env: {
      browser: true,
    },
    extends: [],
    settings: {
      react: {
        version: "detect",
      },
    },
    rules: {
      "react/jsx-uses-react": "off",
      "react/react-in-jsx-scope": "off",
    },
  };

  const jest = {
    testEnvironment: "jsdom",
  };

  const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
  const content = PackageJsonMerge.merge(packageJson.getContent(), {
    eslintConfig,
    jest,
  });

  const extendsPlugin = "plugin:react/recommended";
  const eslintConfigExtends = (content?.eslintConfig as { extends: string[] }).extends;

  const prettierIndex = eslintConfigExtends.indexOf("prettier");

  if (prettierIndex >= 0) {
    eslintConfigExtends.splice(prettierIndex, 0, extendsPlugin);
  } else {
    eslintConfigExtends.push(extendsPlugin);
  }

  packageJson.setContent(content);
}
