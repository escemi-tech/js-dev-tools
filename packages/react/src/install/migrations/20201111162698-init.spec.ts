import { readPackageJson } from "@ts-dev-tools/core/dist/services/packageJson";
import {
  createTestProjectDir,
  removeTestProjectDir,
  restorePackageJson,
} from "@ts-dev-tools/core/dist/tests/utils";

import { up } from "./20201111162698-init";

describe("Migration 20201111162698-init", () => {
  let testProjectDir: string;

  beforeAll(() => {
    testProjectDir = createTestProjectDir(__filename);
  });

  afterAll(() => {
    removeTestProjectDir(__filename);
  });

  describe("Up", () => {
    afterEach(() => {
      restorePackageJson(__filename);
    });

    it("should apply migration", () => {
      up(testProjectDir);

      const packageJson = readPackageJson(testProjectDir);

      expect(packageJson).toMatchObject({
        eslintConfig: {},
        jest: {},
      });
    });
  });
});
