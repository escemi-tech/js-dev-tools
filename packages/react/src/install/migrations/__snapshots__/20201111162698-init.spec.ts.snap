// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migration 20201111162698-init Up should apply migration 1`] = `
Object {
  "eslintConfig": Object {
    "env": Object {
      "browser": true,
    },
    "extends": Array [
      "plugin:react/recommended",
    ],
    "rules": Object {
      "react/jsx-uses-react": "off",
      "react/react-in-jsx-scope": "off",
    },
    "settings": Object {
      "react": Object {
        "version": "detect",
      },
    },
  },
  "jest": Object {
    "testEnvironment": "jsdom",
  },
  "version": "1.0.0",
}
`;

exports[`Migration 20201111162698-init Up should apply migration after core 1`] = `
Object {
  "eslintConfig": Object {
    "env": Object {
      "browser": true,
      "es2021": true,
    },
    "extends": Array [
      "eslint:recommended",
      "plugin:@typescript-eslint/recommended",
      "plugin:jest/recommended",
      "plugin:react/recommended",
      "prettier",
    ],
    "rules": Object {
      "react/jsx-uses-react": "off",
      "react/react-in-jsx-scope": "off",
    },
    "settings": Object {
      "react": Object {
        "version": "detect",
      },
    },
  },
  "jest": Object {
    "preset": "ts-jest",
    "testEnvironment": "jsdom",
  },
  "version": "1.0.0",
}
`;
